### PR TITLE
feat(ConstTypeRequiredFixer): Introduction

### PR DESCRIPTION
**Pull request by SpacePossum**

### CODE REVIEW

**Comment by SpacePossum** [comment]
rebase on latest, no other changes

**Comment by SpacePossum** [comment]
the failure is odd, it works on the RC so I think this CI is not running it, also fails on 3v4l which states it runs on git-master of php-src IIUC but than this shouldn't fail, anyways, lets see all env. get properly updated to 8.3

File: src/Fixer/ClassNotation/ConstTypeRequiredFixer.php [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+     * Must run after NoUnneededControlParenthesesFixer, NoUselessConcatOperatorFixer.
+     */
+    public function getPriority(): int
+    {
+        return 0;
+    }
+
+    public function isCandidate(Tokens $tokens): bool
+    {
+        return \PHP_VERSION_ID >= 8_03_00
```
Comment (keradus) [inline src/Fixer/ClassNotation/ConstTypeRequiredFixer.php:59]: perhaps add to php 8.3 rulesset?

File: src/Fixer/ClassNotation/ConstTypeRequiredFixer.php [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+
+    private function getTypeForStringToken(Token $token): ?string
+    {
+        \assert($token->isGivenKind(T_STRING));
+
+        $content = $token->getContent();
+        $contentLower = strtolower($content);
+
+        if ('true' === $contentLower || 'false' === $contentLower) {
+            $type = 'bool';
```
Comment (keradus) [inline src/Fixer/ClassNotation/ConstTypeRequiredFixer.php:230]:
```suggestion
        if ('true' === $contentLower) {
            $type = 'true';
        } elseif ('false' === $contentLower) {
            $type = false;
```

File: src/Fixer/ClassNotation/ConstTypeRequiredFixer.php [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+        } elseif ('null' === $contentLower) {
+            $type = 'null';
```
Comment (keradus) [inline src/Fixer/ClassNotation/ConstTypeRequiredFixer.php:232]:
```suggestion
        if ('true' === $contentLower || 'false' === $contentLower || 'null' === $contentLower) {
            $type = $contentLower;
```

File: src/Fixer/ClassNotation/ConstTypeRequiredFixer.php [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+                    new Token([T_WHITESPACE, ' ']),
+                    new Token(['array' === $type ? CT::T_ARRAY_TYPEHINT : T_STRING, $type]),
+                ],
+            ],
+        );
+    }
+
+    private function getTypeForStringToken(Token $token): ?string
+    {
+        \assert($token->isGivenKind(T_STRING));
```
Comment (keradus) [inline src/Fixer/ClassNotation/ConstTypeRequiredFixer.php:224]: ?

File: src/Fixer/ClassNotation/ConstTypeRequiredFixer.php [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+        ];
+
+        return $token->isGivenKind($magicConstStr);
+    }
+
+    private function isPredefinedConstStr(string $content): bool
+    {
+        static $predefinedConstStr = null;
+
+        if (null === $predefinedConstStr) {
```
Comment (keradus) [inline src/Fixer/ClassNotation/ConstTypeRequiredFixer.php:272]: any need for this `if` ?

Comment (SpacePossum) [reply src/Fixer/ClassNotation/ConstTypeRequiredFixer.php:272]: to only create the list:
- when needed @ runtime
- create only max one time

File: src/Fixer/ClassNotation/ConstTypeRequiredFixer.php (hunk 1 above)
Comment (SpacePossum) [reply src/Fixer/ClassNotation/ConstTypeRequiredFixer.php:59]: could be a migrating rule set maybe, however there is no 8.3 rule set yet, maybe best to add new sets in own PR's to keep the PR's smaller?

File: src/Fixer/ClassNotation/ConstTypeRequiredFixer.php (hunk 3 above)
Comment (SpacePossum) [reply src/Fixer/ClassNotation/ConstTypeRequiredFixer.php]: thanks, done in https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7352/commits/cd68b4192a344189951816ecbcba1ccf802018f2

File: src/Fixer/ClassNotation/ConstTypeRequiredFixer.php (hunk 4 above)
Comment (SpacePossum) [reply src/Fixer/ClassNotation/ConstTypeRequiredFixer.php]: removed in https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7352/commits/6a8085fe365c080400f6006d6415ead1e609ee56

**Comment by SpacePossum** [comment]
added the two commits based on review feedback, than rebased on latest and forced pushed

**Comment by SpacePossum** [comment]
redid this is refactor tool, which is better fit as this is a CS tool

**Comment by Wirone** [comment]
> this is refactor tool, which is better fit as this is a CS tool

I believe @keradus' definition of "coding standards" is wider than only formatting 🙂. There are multiple rules closer to refactoring than reformatting already.

File: src/Fixer/ClassNotation/ConstTypeRequiredFixer.php (hunk 1 above)
Comment (keradus) [reply src/Fixer/ClassNotation/ConstTypeRequiredFixer.php:59]: I'm happy with both

**Comment by keradus** [comment]
Coding Standards are definitely more than only whitespaces handling ;) we already move lot of phpdoc to new code syntax [like typing], so definitely one to go with ;)

**Pull request closed without merging**